Readd grpclb fallback

AspirinSJL opened the pull request:
This is a continuation of #12187.

Issues I need to fix:
- @sreecha Test times out. This should be fixed by avoiding multiple fallback timers.
- @dgquintas Memory leakage: https://gist.github.com/dgquintas/42b4d0c7e639d5eea22f7b2ebea030b6.
- @summerxyt Integration test failure. Fallback address update is not working where backends are restarted.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/grpc/grpc/12587)
<!-- Reviewable:end -->

grpc-jenkins commented:
```
[trickle] No significant performance differences
```

dgquintas commented:
This PR is in progress, right? If so, please add the "do not merge" and "work in progress" labels, to indicate us when we can start reviewing.

AspirinSJL commented:
Yes. Just added the labels. Thanks!

grpc-jenkins commented:
```
[microbenchmarks] Performance differences noted:
Benchmark                                                             cpu_time    nows_per_iteration    real_time
--------------------------------------------------------------------  ----------  --------------------  -----------
BM_ClosureRunOnExecCtx                                                +4%                               +4%
BM_UnaryPingPong<MinInProcessCHTTP2, NoOpMutator, NoOpMutator>/0/16M              -5%
```

grpc-jenkins commented:
```
Corrupt JSON data (indicates timeout or crash): 
    bm_fullstack_trickle.BM_PumpStreamServerToClient_Trickle_8_4k.opt.old: 1
    bm_fullstack_trickle.BM_PumpStreamServerToClient_Trickle_64_4k.opt.new: 1


[trickle] No significant performance differences
```

grpc-jenkins commented:
```
[trickle] No significant performance differences
```

grpc-jenkins commented:
```
[microbenchmarks] Performance differences noted:
Benchmark                                                                                cpu_time    real_time
---------------------------------------------------------------------------------------  ----------  -----------
BM_StreamingPingPongWithCoalescingApi<InProcessCHTTP2, NoOpMutator, NoOpMutator>/64/2/0  +4%         +4%
```

grpc-jenkins commented:
```
[trickle] Performance differences noted:
Benchmark                                      svr_transport_stalls_per_iteration
---------------------------------------------  ------------------------------------
BM_PumpStreamServerToClient_Trickle/128M/128M  +9999%
```

grpc-jenkins commented:
```
Corrupt JSON data (indicates timeout or crash): 
    bm_fullstack_streaming_ping_pong.BM_StreamingPingPongWithCoalescingApi_InProcessCHTTP2_NoOpMutator_NoOpMutator__2M_2_0.opt.new: 1


[microbenchmarks] Performance differences noted:
Benchmark                                                                              cpu_time    real_time
-------------------------------------------------------------------------------------  ----------  -----------
BM_PumpStreamClientToServer<InProcess>/32k                                             +9%         +10%
BM_PumpStreamServerToClient<InProcess>/256k                                            +7%         +7%
BM_PumpStreamServerToClient<InProcess>/32k                                             +12%        +12%
BM_PumpStreamServerToClient<InProcessCHTTP2>/32k                                       +9%         +9%
BM_StreamingPingPongMsgs<InProcess, NoOpMutator, NoOpMutator>/256k                     -5%         -5%
BM_StreamingPingPongMsgs<InProcess, NoOpMutator, NoOpMutator>/32k                      -4%
BM_StreamingPingPongMsgs<MinInProcess, NoOpMutator, NoOpMutator>/32k                   -7%         -7%
BM_StreamingPingPongWithCoalescingApi<MinInProcess, NoOpMutator, NoOpMutator>/32k/2/0  -12%        -11%
BM_UnaryPingPong<InProcess, NoOpMutator, NoOpMutator>/0/32k                            +4%         +4%
BM_UnaryPingPong<InProcess, NoOpMutator, NoOpMutator>/256k/0                           +8%         +8%
BM_UnaryPingPong<InProcessCHTTP2, NoOpMutator, NoOpMutator>/256k/0                     +8%         +8%
BM_UnaryPingPong<MinInProcess, NoOpMutator, NoOpMutator>/256k/256k                     +9%         +10%
BM_UnaryPingPong<MinInProcess, NoOpMutator, NoOpMutator>/32k/32k                       +4%         +5%
BM_UnaryPingPong<MinInProcessCHTTP2, NoOpMutator, NoOpMutator>/32k/32k                 +5%         +4%
```

grpc-jenkins commented:
```
[trickle] No significant performance differences
```

grpc-jenkins commented:
```
[trickle] No significant performance differences
```

grpc-jenkins commented:
```
Corrupt JSON data (indicates timeout or crash): 
    bm_fullstack_streaming_ping_pong.BM_StreamingPingPongWithCoalescingApi_MinInProcessCHTTP2_NoOpMutator_NoOpMutator__2M_2_0.counters.new: 1
    bm_fullstack_streaming_ping_pong.BM_StreamingPingPongWithCoalescingApi_MinInProcessCHTTP2_NoOpMutator_NoOpMutator__2M_2_0.counters.old: 1


[microbenchmarks] Performance differences noted:
Benchmark                                                                               atm_add_per_iteration    cpu_time    nows_per_iteration    real_time
--------------------------------------------------------------------------------------  -----------------------  ----------  --------------------  -----------
BM_StreamingPingPong<MinTCP, NoOpMutator, NoOpMutator>/2M/1                             -4%                                  -4%
BM_StreamingPingPongWithCoalescingApi<InProcess, NoOpMutator, NoOpMutator>/4k/2/0                                +4%                               +4%
BM_StreamingPingPongWithCoalescingApi<InProcessCHTTP2, NoOpMutator, NoOpMutator>/8/1/1                           +12%                              +12%
BM_UnaryPingPong<InProcess, NoOpMutator, NoOpMutator>/0/1                                                        +5%                               +5%
```

grpc-jenkins commented:
```
[trickle] No significant performance differences
```

grpc-jenkins commented:
```
[microbenchmarks] Performance differences noted:
Benchmark                                                                                    cpu_time    nows_per_iteration    real_time
-------------------------------------------------------------------------------------------  ----------  --------------------  -----------
BM_StreamingPingPongWithCoalescingApi<InProcess, NoOpMutator, NoOpMutator>/1/1/1             +6%                               +5%
BM_StreamingPingPongWithCoalescingApi<MinInProcessCHTTP2, NoOpMutator, NoOpMutator>/16M/1/1              -5%
BM_TransportStreamSend/2M                                                                                                      +4%
BM_UnaryPingPong<MinInProcess, NoOpMutator, NoOpMutator>/64/64                               +4%                               +4%
BM_UnaryPingPong<MinInProcessCHTTP2, NoOpMutator, NoOpMutator>/32k/32k                       +6%                               +7%
```

AspirinSJL commented:
There is a flake that I don't know how to fix. It happens about 1/1000.

```
FAILED: bins/opt/grpclb_end2end_test --gtest_filter=SingleBalancerTest.FallbackUpdate  GRPC_POLL_STRATEGY=epoll1 [ret=-6, pid=189863]

I0925 16:27:11.936582061    1231 ev_epoll1_linux.c:93]       grpc epoll fd: 3
D0925 16:27:11.936629270    1231 ev_posix.c:107]             Using polling engine: epoll1
D0925 16:27:11.936648500    1231 dns_resolver.c:301]         Using native dns resolver
D0925 16:27:11.936680510    1231 test_config.c:388]          test slowdown factor: sanitizer=1, fixture=1, poller=1, total=1
Note: Google Test filter = SingleBalancerTest.FallbackUpdate
[==========] Running 1 test from 1 test case.
[----------] Global test environment set-up.
[----------] 1 test from SingleBalancerTest
[ RUN      ] SingleBalancerTest.FallbackUpdate
I0925 16:27:11.938828392    1231 grpclb_end2end_test.cc:525] starting backend server on port 2241
I0925 16:27:11.938985336    1264 server_builder.cc:269]      Synchronous server. Num CQs: 1, Min pollers: 1, Max Pollers: 2, CQ timeout (msec): 10000
I0925 16:27:11.940071604    1231 grpclb_end2end_test.cc:530] backend server startup complete
I0925 16:27:11.942032965    1231 grpclb_end2end_test.cc:525] starting backend server on port 1116
I0925 16:27:11.942095100    1287 server_builder.cc:269]      Synchronous server. Num CQs: 1, Min pollers: 1, Max Pollers: 2, CQ timeout (msec): 10000
I0925 16:27:11.943220799    1231 grpclb_end2end_test.cc:530] backend server startup complete
I0925 16:27:11.944276360    1231 grpclb_end2end_test.cc:525] starting backend server on port 7487
I0925 16:27:11.944349386    1298 server_builder.cc:269]      Synchronous server. Num CQs: 1, Min pollers: 1, Max Pollers: 2, CQ timeout (msec): 10000
I0925 16:27:11.945420614    1231 grpclb_end2end_test.cc:530] backend server startup complete
I0925 16:27:11.946173295    1231 grpclb_end2end_test.cc:525] starting backend server on port 14855
I0925 16:27:11.946219698    1311 server_builder.cc:269]      Synchronous server. Num CQs: 1, Min pollers: 1, Max Pollers: 2, CQ timeout (msec): 10000
I0925 16:27:11.947494103    1231 grpclb_end2end_test.cc:530] backend server startup complete
I0925 16:27:11.948205623    1231 grpclb_end2end_test.cc:525] starting balancer server on port 13202
I0925 16:27:11.948262070    1317 server_builder.cc:269]      Synchronous server. Num CQs: 1, Min pollers: 1, Max Pollers: 2, CQ timeout (msec): 10000
I0925 16:27:11.948995609    1231 grpclb_end2end_test.cc:530] balancer server startup complete
I0925 16:27:11.949422822    1321 grpclb_end2end_test.cc:188] LB[0x1a3f2e0]: BalanceLoad
I0925 16:27:11.949554969    1321 grpclb_end2end_test.cc:192] LB[0x1a3f2e0]: recv msg 'initial_request {
  name: "servername_not_used"
}
'
I0925 16:27:11.949563022    1321 grpclb_end2end_test.cc:304] LB[0x1a3f2e0]: sleeping for 1000 ms...
I0925 16:27:12.150199662    1231 grpclb_end2end_test.cc:738] ========= BEFORE FIRST BATCH ==========
I0925 16:27:12.150370177    1231 grpclb_end2end_test.cc:740] ========= DONE WITH FIRST BATCH ==========
I0925 16:27:12.151234826    1231 grpclb_end2end_test.cc:765] ========= BEFORE SECOND BATCH ==========
I0925 16:27:12.151369424    1231 grpclb_end2end_test.cc:767] ========= DONE WITH SECOND BATCH ==========
I0925 16:27:12.949654479    1321 grpclb_end2end_test.cc:310] LB[0x1a3f2e0]: Woke up! Sending response 'server_list {
  servers {
    ip_address: "\177\000\000\001"
    port: 7487
  }
  servers {
    ip_address: "\177\000\000\001"
    port: 14855
  }
}
'
I0925 16:27:12.949737264    1441 grpclb.c:539]               Missing LB token for backend address 'ipv4:127.0.0.1:7487'. The empty token will be used instead
I0925 16:27:12.949754861    1441 grpclb.c:539]               Missing LB token for backend address 'ipv4:127.0.0.1:14855'. The empty token will be used instead
I0925 16:27:12.950866925    1231 grpclb_end2end_test.cc:789] ========= BEFORE THIRD BATCH ==========
I0925 16:27:12.951448621    1231 grpclb_end2end_test.cc:792] ========= DONE WITH THIRD BATCH ==========
I0925 16:27:12.951465234    1231 grpclb_end2end_test.cc:141] Backend: shut down
I0925 16:27:12.951471213    1231 grpclb_end2end_test.cc:549] backend about to shutdown
I0925 16:27:12.951552713    1231 grpclb_end2end_test.cc:552] backend shutdown completed
I0925 16:27:12.951568898    1231 grpclb_end2end_test.cc:141] Backend: shut down
I0925 16:27:12.951573802    1231 grpclb_end2end_test.cc:549] backend about to shutdown
I0925 16:27:12.951663653    1231 grpclb_end2end_test.cc:552] backend shutdown completed
I0925 16:27:12.951669885    1231 grpclb_end2end_test.cc:141] Backend: shut down
I0925 16:27:12.951672281    1231 grpclb_end2end_test.cc:549] backend about to shutdown
I0925 16:27:12.951833650    1231 grpclb_end2end_test.cc:552] backend shutdown completed
I0925 16:27:12.951840506    1231 grpclb_end2end_test.cc:141] Backend: shut down
I0925 16:27:12.951842867    1231 grpclb_end2end_test.cc:549] backend about to shutdown
I0925 16:27:12.951968292    1231 grpclb_end2end_test.cc:552] backend shutdown completed
I0925 16:27:12.951975099    1231 grpclb_end2end_test.cc:265] LB[0x1a3f2e0]: shut down
I0925 16:27:12.951977462    1231 grpclb_end2end_test.cc:549] balancer about to shutdown
E0925 16:27:12.952106550    2135 grpclb.c:1876]              assertion failed: glb_policy->lb_call == NULL



*******************************
Caught signal SIGABRT
I0925 16:27:12.952132151    1321 grpclb_end2end_test.cc:250] LB[0x1a3f2e0]: done
bins/opt/grpclb_end2end_test[0x5901ee]
/lib/x86_64-linux-gnu/libpthread.so.0(+0x10330)[0x7f74535e1330]
/lib/x86_64-linux-gnu/libc.so.6(gsignal+0x37)[0x7f7452805c37]
/lib/x86_64-linux-gnu/libc.so.6(abort+0x148)[0x7f7452809028]
bins/opt/grpclb_end2end_test[0x4cbea0]
bins/opt/grpclb_end2end_test[0x44f726]
bins/opt/grpclb_end2end_test[0x451c78]
bins/opt/grpclb_end2end_test[0x462fdc]
bins/opt/grpclb_end2end_test[0x48d294]
bins/opt/grpclb_end2end_test[0x48d72b]
bins/opt/grpclb_end2end_test[0x4707a8]
bins/opt/grpclb_end2end_test[0x476e5d]
bins/opt/grpclb_end2end_test[0x47d3a5]
bins/opt/grpclb_end2end_test[0x47d4cc]
/usr/lib/x86_64-linux-gnu/libstdc++.so.6(+0xb1a60)[0x7f7453165a60]
/lib/x86_64-linux-gnu/libpthread.so.0(+0x8184)[0x7f74535d9184]
/lib/x86_64-linux-gnu/libc.so.6(clone+0x6d)[0x7f74528ccffd]
```

It only happens in `SingleBalancerTest.FallbackUpdate`, also the pre-existing `UpdatesTest.UpdateBalancersRepeated` has the same flake stacktrace, so I think they have the same root cause. @dgquintas can check.

Otherwise, I don't see directly-related issues to fix.

grpc-jenkins commented:
```
[trickle] No significant performance differences
```

grpc-jenkins commented:
```
Corrupt JSON data (indicates timeout or crash): 
    bm_fullstack_streaming_ping_pong.BM_StreamingPingPongWithCoalescingApi_MinInProcessCHTTP2_NoOpMutator_NoOpMutator__2M_2_0.opt.old: 1


[microbenchmarks] Performance differences noted:
Benchmark                                                                                                    atm_cas_per_iteration    cpu_time    real_time
-----------------------------------------------------------------------------------------------------------  -----------------------  ----------  -----------
BM_CreateDestroyPollset                                                                                                               +6%         +6%
BM_ErrorStringOnNewError<SimpleError>                                                                                                 +6%         +6%
BM_HpackEncoderEncodeHeader<EmptyBatch>/0/16k                                                                                         +4%
BM_HpackEncoderEncodeHeader<SingleInternedBinaryElem<1, false>>/0/16k                                                                 +6%         +6%
BM_HpackEncoderEncodeHeader<SingleNonInternedBinaryElem<100, false>>/0/16k                                                            +4%         +4%
BM_HpackParserParseHeader<IndexedSingleStaticElem>                                                                                    +6%         +6%
BM_Pluck1Core                                                                                                                         +5%         +6%
BM_PumpStreamClientToServer<InProcessCHTTP2>/32k                                                                                      +8%         +8%
BM_PumpStreamClientToServer<SockPair>/512                                                                                                         +5%
BM_PumpStreamServerToClient<InProcess>/0                                                                                                          +4%
BM_PumpStreamServerToClient<SockPair>/64                                                                                                          +7%
BM_StreamingPingPong<InProcessCHTTP2, NoOpMutator, NoOpMutator>/512/1                                                                 +7%         +7%
BM_StreamingPingPong<MinInProcessCHTTP2, NoOpMutator, NoOpMutator>/64/1                                                               +18%        +17%
BM_StreamingPingPong<TCP, NoOpMutator, NoOpMutator>/0/1                                                                               +7%         +7%
BM_StreamingPingPongWithCoalescingApi<InProcess, NoOpMutator, NoOpMutator>/4k/2/1                                                     +12%        +12%
BM_StreamingPingPongWithCoalescingApi<InProcess, NoOpMutator, NoOpMutator>/512/2/0                                                    +5%         +6%
BM_StreamingPingPongWithCoalescingApi<InProcess, NoOpMutator, NoOpMutator>/8/2/1                                                      +14%        +14%
BM_StreamingPingPongWithCoalescingApi<MinInProcess, NoOpMutator, NoOpMutator>/32k/1/1                                                 +4%         +4%
BM_StreamingPingPongWithCoalescingApi<MinInProcess, NoOpMutator, NoOpMutator>/4k/2/1                                                  +4%         +4%
BM_StreamingPingPongWithCoalescingApi<MinInProcess, NoOpMutator, NoOpMutator>/512/2/0                                                 +9%         +9%
BM_UnaryPingPong<InProcess, Client_AddMetadata<RandomBinaryMetadata<10>, 2>, NoOpMutator>/0/0                                         +5%         +5%
BM_UnaryPingPong<InProcess, NoOpMutator, NoOpMutator>/0/0                                                                             +8%         +8%
BM_UnaryPingPong<InProcess, NoOpMutator, NoOpMutator>/0/2M                                                   +5%
BM_UnaryPingPong<InProcess, NoOpMutator, NoOpMutator>/512/0                                                                           +10%        +9%
BM_UnaryPingPong<InProcess, NoOpMutator, NoOpMutator>/512/512                                                                         +5%         +5%
BM_UnaryPingPong<InProcess, NoOpMutator, Server_AddInitialMetadata<RandomAsciiMetadata<100>, 1>>/0/0                                  +5%         +5%
BM_UnaryPingPong<InProcessCHTTP2, Client_AddMetadata<RandomBinaryMetadata<31>, 1>, NoOpMutator>/0/0                                   +4%         +4%
BM_UnaryPingPong<InProcessCHTTP2, NoOpMutator, NoOpMutator>/4k/4k                                                                     +15%        +15%
BM_UnaryPingPong<InProcessCHTTP2, NoOpMutator, NoOpMutator>/512/0                                                                     +4%         +4%
BM_UnaryPingPong<InProcessCHTTP2, NoOpMutator, Server_AddInitialMetadata<RandomAsciiMetadata<10>, 100>>/0/0                           +4%         +4%
BM_UnaryPingPong<MinInProcessCHTTP2, NoOpMutator, NoOpMutator>/0/4k                                                                   +4%         +5%
BM_UnaryPingPong<MinInProcessCHTTP2, NoOpMutator, NoOpMutator>/0/8                                                                    +7%         +7%
BM_UnaryPingPong<MinInProcessCHTTP2, NoOpMutator, NoOpMutator>/1/1                                                                    +5%         +5%
BM_UnaryPingPong<TCP, NoOpMutator, NoOpMutator>/4k/0                                                                                  +5%         +5%
```

markdroth commented:
Does the flake you're seeing still happen in `UpdatesTest.UpdateBalancersRepeated` even without this PR?  If so, then I agree that it's probably unrelated to this PR, but @dgquintas should verify.

Overall, the code looks good.  My comments are mostly minor.

---

Reviewed 5 of 7 files at r1, 1 of 1 files at r5, 1 of 1 files at r6.
Review status: all files reviewed at latest revision, 8 unresolved discussions, some commit checks broke.

---

*[src/core/ext/filters/client_channel/lb_policy/grpclb/grpclb.c, line 1678 at r6](https://reviewable.io:443/reviews/grpc/grpc/12587#-KuycolMF0gIyOi_3qIv:-KuycolMF0gIyOi_3qIw:b-9f2ltu) ([raw file](https://github.com/grpc/grpc/blob/373d0783d37946586c4b127032191fa2a93f98b4/src/core/ext/filters/client_channel/lb_policy/grpclb/grpclb.c#L1678)):*
> ```C
>    * actually runs, don't do anything. */
>   if (glb_policy->serverlist != NULL) {
>     GRPC_LB_POLICY_WEAK_UNREF(exec_ctx, &glb_policy->base,
> ```

This unref already happens at the end of the function, so we could eliminate the duplication by inverting the conditional and moving the rest of the function inside of it:

```
if (glb_policy->serverlist == NULL) {
  // ... the existing code from lines 1683-1691 ...
}
GRPC_LB_POLICY_WEAK_UNREF(exec_ctx, &glb_policy->base,
                          "grpclb_fallback_timer");
```

---

*[src/core/ext/filters/client_channel/lb_policy/grpclb/grpclb.c, line 1682 at r6](https://reviewable.io:443/reviews/grpc/grpc/12587#-KuycSvYop1bxeJQAsVA:-KuycSvZb_PYOe5njtIw:bxx4mc1) ([raw file](https://github.com/grpc/grpc/blob/373d0783d37946586c4b127032191fa2a93f98b4/src/core/ext/filters/client_channel/lb_policy/grpclb/grpclb.c#L1682)):*
> ```C
>     return;
>   }
>   glb_policy->fallback_timer_active = false;
> ```

Shouldn't we do this even if `glb_policy->serverlist` is null?  Might want to move this above the preceding conditional.

---

*[src/core/ext/filters/client_channel/lb_policy/grpclb/grpclb.c, line 1750 at r6](https://reviewable.io:443/reviews/grpc/grpc/12587#-KuygwkF1hXBbxKmmQvZ:-KuygwkF1hXBbxKmmQv_:b-v1beo4) ([raw file](https://github.com/grpc/grpc/blob/373d0783d37946586c4b127032191fa2a93f98b4/src/core/ext/filters/client_channel/lb_policy/grpclb/grpclb.c#L1750)):*
>   const grpc_arg *arg =
>       grpc_channel_args_find(args->args, GRPC_ARG_LB_ADDRESSES);
>   if (arg == NULL || arg->type != GRPC_ARG_POINTER) {
>     return;
>   }
>   const grpc_lb_addresses *addresses =
>       (const grpc_lb_addresses *)arg->value.pointer.p;

This code duplicates what already exists in `glb_update_locked()`.  To avoid this duplication, I suggest the following:

- Move the code from lines 1798-1818 to the top of `glb_update_locked()`, so that we have already constructed the `grpc_lb_addresses` object before we call this function.
- Pass in the `grpc_lb_addresses` object as a parameter to this function.

---

*[src/core/ext/filters/client_channel/lb_policy/grpclb/grpclb.c, line 1761 at r6](https://reviewable.io:443/reviews/grpc/grpc/12587#-KuyfseBp8Uhl5-1W9CS:-KuyfseBp8Uhl5-1W9CT:b7mk2rh) ([raw file](https://github.com/grpc/grpc/blob/373d0783d37946586c4b127032191fa2a93f98b4/src/core/ext/filters/client_channel/lb_policy/grpclb/grpclb.c#L1761)):*
> ```C
>   glb_policy->fallback_backend_addresses =
>       extract_backend_addresses_locked(exec_ctx, addresses);
>   if (glb_policy->lb_fallback_timeout_ms > 0 &&
> ```

This was originally `if (glb_policy->rr_policy != NULL)`.  Why change it?  What was wrong with the original condition?

---

*[src/core/ext/filters/client_channel/lb_policy/grpclb/grpclb.c, line 1776 at r6](https://reviewable.io:443/reviews/grpc/grpc/12587#-KuyfaT_ZdnL2vQEBsZt:-KuyfaT_ZdnL2vQEBsZu:bkyccwk) ([raw file](https://github.com/grpc/grpc/blob/373d0783d37946586c4b127032191fa2a93f98b4/src/core/ext/filters/client_channel/lb_policy/grpclb/grpclb.c#L1776)):*
> ```C
>   }
> 
>   if (glb_policy->updating_lb_channel && glb_policy->serverlist != NULL) {
> ```

This could just be:

```
else if (glb_policy->updating_lb_channel)
```

---

*[test/cpp/end2end/grpclb_end2end_test.cc, line 684 at r6](https://reviewable.io:443/reviews/grpc/grpc/12587#-Kuyi-ESxLjY6IC6VZl1:-Kuyi-ESxLjY6IC6VZl2:b4lmtbx) ([raw file](https://github.com/grpc/grpc/blob/373d0783d37946586c4b127032191fa2a93f98b4/test/cpp/end2end/grpclb_end2end_test.cc#L684)):*
> ```cpp
>   }
> 
>   // Wait until the serverlist reception has been processed.
> ```

...and all backends are reachable.

---

*[test/cpp/end2end/grpclb_end2end_test.cc, line 759 at r6](https://reviewable.io:443/reviews/grpc/grpc/12587#-Kuyi4F1f_vGbNo-JiFc:-Kuyi4F1f_vGbNo-JiFd:b4lmtbx) ([raw file](https://github.com/grpc/grpc/blob/373d0783d37946586c4b127032191fa2a93f98b4/test/cpp/end2end/grpclb_end2end_test.cc#L759)):*
> ```cpp
>   SetNextResolution(addresses);
> 
>   // Wait until the resolution update has been processed.
> ```

...and all backends are reachable.

---

*[test/cpp/end2end/grpclb_end2end_test.cc, line 784 at r6](https://reviewable.io:443/reviews/grpc/grpc/12587#-Kuyi6x-tTlRKk7RKtP-:-Kuyi6x-tTlRKk7RKtP0:b4lmtbx) ([raw file](https://github.com/grpc/grpc/blob/373d0783d37946586c4b127032191fa2a93f98b4/test/cpp/end2end/grpclb_end2end_test.cc#L784)):*
> ```cpp
>   }
> 
>   // Wait until the serverlist reception has been processed.
> ```

...and all backends are reachable.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/grpc/grpc/12587#-:-Kuyjmi7YdPks-NpnVX5:b-ld3kgx)*
<!-- Sent from Reviewable.io -->

grpc-jenkins commented:
```
[trickle] No significant performance differences
```

AspirinSJL commented:
Yes, the flake in `UpdatesTest.UpdateBalancersRepeated` happens at the master branch too, with similar flakiness.  

---

Review status: all files reviewed at latest revision, 8 unresolved discussions, some commit checks broke.

---

*[src/core/ext/filters/client_channel/lb_policy/grpclb/grpclb.c, line 1678 at r6](https://reviewable.io:443/reviews/grpc/grpc/12587#-KuycolMF0gIyOi_3qIv:-KuyzQFhh92IMy7O2aHb:b-896fix) ([raw file](https://github.com/grpc/grpc/blob/373d0783d37946586c4b127032191fa2a93f98b4/src/core/ext/filters/client_channel/lb_policy/grpclb/grpclb.c#L1678)):*
<details><summary><i>Previously, markdroth (Mark D. Roth) wrote…</i></summary><blockquote>

This unref already happens at the end of the function, so we could eliminate the duplication by inverting the conditional and moving the rest of the function inside of it:

```
if (glb_policy->serverlist == NULL) {
  // ... the existing code from lines 1683-1691 ...
}
GRPC_LB_POLICY_WEAK_UNREF(exec_ctx, &glb_policy->base,
                          "grpclb_fallback_timer");
```
</blockquote></details>

Done.

---

*[src/core/ext/filters/client_channel/lb_policy/grpclb/grpclb.c, line 1682 at r6](https://reviewable.io:443/reviews/grpc/grpc/12587#-KuycSvYop1bxeJQAsVA:-Kuz0R5HjuprgqXAE-eL:b-896fix) ([raw file](https://github.com/grpc/grpc/blob/373d0783d37946586c4b127032191fa2a93f98b4/src/core/ext/filters/client_channel/lb_policy/grpclb/grpclb.c#L1682)):*
<details><summary><i>Previously, markdroth (Mark D. Roth) wrote…</i></summary><blockquote>

Shouldn't we do this even if `glb_policy->serverlist` is null?  Might want to move this above the preceding conditional.
</blockquote></details>

Done.

---

*[src/core/ext/filters/client_channel/lb_policy/grpclb/grpclb.c, line 1750 at r6](https://reviewable.io:443/reviews/grpc/grpc/12587#-KuygwkF1hXBbxKmmQvZ:-Kuz6R_CuQuXOv2IOmws:b-896fix) ([raw file](https://github.com/grpc/grpc/blob/373d0783d37946586c4b127032191fa2a93f98b4/src/core/ext/filters/client_channel/lb_policy/grpclb/grpclb.c#L1750)):*
<details><summary><i>Previously, markdroth (Mark D. Roth) wrote…</i></summary><blockquote>

>   const grpc_arg *arg =
>       grpc_channel_args_find(args->args, GRPC_ARG_LB_ADDRESSES);
>   if (arg == NULL || arg->type != GRPC_ARG_POINTER) {
>     return;
>   }
>   const grpc_lb_addresses *addresses =
>       (const grpc_lb_addresses *)arg->value.pointer.p;

This code duplicates what already exists in `glb_update_locked()`.  To avoid this duplication, I suggest the following:

- Move the code from lines 1798-1818 to the top of `glb_update_locked()`, so that we have already constructed the `grpc_lb_addresses` object before we call this function.
- Pass in the `grpc_lb_addresses` object as a parameter to this function.
</blockquote></details>

Done.

---

*[src/core/ext/filters/client_channel/lb_policy/grpclb/grpclb.c, line 1761 at r6](https://reviewable.io:443/reviews/grpc/grpc/12587#-KuyfseBp8Uhl5-1W9CS:-Kuz8FeRmkthwfYbODBs:b-hnoyx3) ([raw file](https://github.com/grpc/grpc/blob/373d0783d37946586c4b127032191fa2a93f98b4/src/core/ext/filters/client_channel/lb_policy/grpclb/grpclb.c#L1761)):*
<details><summary><i>Previously, markdroth (Mark D. Roth) wrote…</i></summary><blockquote>

This was originally `if (glb_policy->rr_policy != NULL)`.  Why change it?  What was wrong with the original condition?
</blockquote></details>

The problem of the previous code is that it doesn't include the case when `glb_policy->rr_policy == NULL` and we do want to propagate the update. For example, the fallback backends are restarted so there is a period without valid `rr_policy`. We saw this situation in our integration test.

---

*[src/core/ext/filters/client_channel/lb_policy/grpclb/grpclb.c, line 1776 at r6](https://reviewable.io:443/reviews/grpc/grpc/12587#-KuyfaT_ZdnL2vQEBsZt:-KuzCRXJVPJA98LAmYfw:bjikzaa) ([raw file](https://github.com/grpc/grpc/blob/373d0783d37946586c4b127032191fa2a93f98b4/src/core/ext/filters/client_channel/lb_policy/grpclb/grpclb.c#L1776)):*
<details><summary><i>Previously, markdroth (Mark D. Roth) wrote…</i></summary><blockquote>

This could just be:

```
else if (glb_policy->updating_lb_channel)
```
</blockquote></details>

Done.

BTW, `&& glb_policy->serverlist != NULL` is added in this line to overwrite the bogus balancer update which takes forever in the integration test.

---

*[test/cpp/end2end/grpclb_end2end_test.cc, line 684 at r6](https://reviewable.io:443/reviews/grpc/grpc/12587#-Kuyi-ESxLjY6IC6VZl1:-KuzEHbb3njes_kdpn_e:bb8jfvo) ([raw file](https://github.com/grpc/grpc/blob/373d0783d37946586c4b127032191fa2a93f98b4/test/cpp/end2end/grpclb_end2end_test.cc#L684)):*
<details><summary><i>Previously, markdroth (Mark D. Roth) wrote…</i></summary><blockquote>

...and all backends are reachable.
</blockquote></details>

Actually, if I add `WaitForAllBackends()` at the beginning of this test just like other tests in this file, this test always fails because it waits until the `serverlist` is received. That's why I separated all the waiting mutual-exclusively.

But now I wonder why `WaitForAllBackends()` can finish in the fallback related test cases.. 

I wanted to change the waiting into a incremental way. All the increments don't introduce problems (which looks strange to me now..) except the `serverlist` reception part in `FallbackUpdate` test.

---

*[test/cpp/end2end/grpclb_end2end_test.cc, line 759 at r6](https://reviewable.io:443/reviews/grpc/grpc/12587#-Kuyi4F1f_vGbNo-JiFc:-KuzG2EjJqE2LT7bHZ_0:bb9xe20) ([raw file](https://github.com/grpc/grpc/blob/373d0783d37946586c4b127032191fa2a93f98b4/test/cpp/end2end/grpclb_end2end_test.cc#L759)):*
<details><summary><i>Previously, markdroth (Mark D. Roth) wrote…</i></summary><blockquote>

...and all backends are reachable.
</blockquote></details>

The same as above (line 684).

---

*[test/cpp/end2end/grpclb_end2end_test.cc, line 784 at r6](https://reviewable.io:443/reviews/grpc/grpc/12587#-Kuyi6x-tTlRKk7RKtP-:-KuzG7gOWaF2gfOLfgOY:bb9xe20) ([raw file](https://github.com/grpc/grpc/blob/373d0783d37946586c4b127032191fa2a93f98b4/test/cpp/end2end/grpclb_end2end_test.cc#L784)):*
<details><summary><i>Previously, markdroth (Mark D. Roth) wrote…</i></summary><blockquote>

...and all backends are reachable.
</blockquote></details>

The same as above (line 684).

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/grpc/grpc/12587#-:-KuywORFkrupAQm5ZBKV:b-sig8b9)*
<!-- Sent from Reviewable.io -->

dgquintas commented:
I'm having a look at the preexisting flake in `UpdatesTest.UpdateBalancersRepeated`

grpc-jenkins commented:
```
Corrupt JSON data (indicates timeout or crash): 
    bm_fullstack_streaming_ping_pong.BM_StreamingPingPongWithCoalescingApi_MinInProcessCHTTP2_NoOpMutator_NoOpMutator__2M_2_0.counters.old: 1


[microbenchmarks] Performance differences noted:
Benchmark                                                                             cpu_time    locks_per_iteration    nows_per_iteration    real_time
------------------------------------------------------------------------------------  ----------  ---------------------  --------------------  -----------
BM_HpackEncoderEncodeHeader<SingleNonInternedBinaryElem<100, true>>/0/16k             +8%                                                      +9%
BM_PumpStreamClientToServer<InProcess>/512                                            +6%                                                      +6%
BM_PumpStreamServerToClient<UDS>/512                                                                                                           +5%
BM_SliceInternEqualToStaticMetadata                                                   +4%                                                      +4%
BM_StreamingPingPongMsgs<InProcess, NoOpMutator, NoOpMutator>/512                     +5%                                                      +5%
BM_StreamingPingPongMsgs<InProcess, NoOpMutator, NoOpMutator>/64                                                                               +4%
BM_StreamingPingPongWithCoalescingApi<MinInProcess, NoOpMutator, NoOpMutator>/64/2/1  +4%                                                      +4%
BM_TransportStreamRecv/8                                                              +4%                                                      +4%
BM_UnaryPingPong<MinInProcess, NoOpMutator, NoOpMutator>/128M/0                                   +6%                    +6%
BM_UnaryPingPong<MinInProcessCHTTP2, NoOpMutator, NoOpMutator>/4k/4k                  +12%                                                     +13%
```

dgquintas commented:
*[test/cpp/end2end/grpclb_end2end_test.cc, line 684 at r6](https://reviewable.io:443/reviews/grpc/grpc/12587#-Kuyi-ESxLjY6IC6VZl1:-KuzaspBNFZR7H5JKje2:b-ypvpft) ([raw file](https://github.com/grpc/grpc/blob/373d0783d37946586c4b127032191fa2a93f98b4/test/cpp/end2end/grpclb_end2end_test.cc#L684)):*
<details><summary><i>Previously, AspirinSJL (Juanli Shen) wrote…</i></summary><blockquote>

Actually, if I add `WaitForAllBackends()` at the beginning of this test just like other tests in this file, this test always fails because it waits until the `serverlist` is received. That's why I separated all the waiting mutual-exclusively.

But now I wonder why `WaitForAllBackends()` can finish in the fallback related test cases.. 

I wanted to change the waiting into a incremental way. All the increments don't introduce problems (which looks strange to me now..) except the `serverlist` reception part in `FallbackUpdate` test.

</blockquote></details>

`WaitForAllBackends()` simply waits until all backends have serviced a request, it knows nothing about serverlists. Because in the fallback case you are hitting the backends directly, it's able to make progress. In the general case (without fallback active), it's true that a serverlist is a prerequisite, but only because that's the only way grpclb has to find what the backends are.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/grpc/grpc/12587)*
<!-- Sent from Reviewable.io -->

AspirinSJL commented:
Review status: 6 of 7 files reviewed at latest revision, 8 unresolved discussions, some commit checks broke.

---

*[test/cpp/end2end/grpclb_end2end_test.cc, line 684 at r6](https://reviewable.io:443/reviews/grpc/grpc/12587#-Kuyi-ESxLjY6IC6VZl1:-KuzdhCmOmPRKstByKpb:b-60trpm) ([raw file](https://github.com/grpc/grpc/blob/373d0783d37946586c4b127032191fa2a93f98b4/test/cpp/end2end/grpclb_end2end_test.cc#L684)):*
<details><summary><i>Previously, dgquintas (David G. Quintas) wrote…</i></summary><blockquote>

`WaitForAllBackends()` simply waits until all backends have serviced a request, it knows nothing about serverlists. Because in the fallback case you are hitting the backends directly, it's able to make progress. In the general case (without fallback active), it's true that a serverlist is a prerequisite, but only because that's the only way grpclb has to find what the backends are.
</blockquote></details>

I see. I just realized that I didn't reset the counters after each batch..  Then all of it makes sense to me now.. I feel this PR is mergable soon!!!

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/grpc/grpc/12587)*
<!-- Sent from Reviewable.io -->

AspirinSJL commented:
Review status: 6 of 7 files reviewed at latest revision, 8 unresolved discussions, some commit checks broke.

---

*[test/cpp/end2end/grpclb_end2end_test.cc, line 684 at r6](https://reviewable.io:443/reviews/grpc/grpc/12587#-Kuyi-ESxLjY6IC6VZl1:-Kuzi3oQaC8gSGqlwy_C:bj72j57) ([raw file](https://github.com/grpc/grpc/blob/373d0783d37946586c4b127032191fa2a93f98b4/test/cpp/end2end/grpclb_end2end_test.cc#L684)):*
<details><summary><i>Previously, AspirinSJL (Juanli Shen) wrote…</i></summary><blockquote>

I see. I just realized that I didn't reset the counters after each batch..  Then all of it makes sense to me now.. I feel this PR is mergable soon!!!
</blockquote></details>

I think we should use the previous comments, because `WaitForBackend()` works in a customer view instead of infrastructure view. If `WaitForBackend(i)` finishes, it only means backend `i` is reachable (from the client).

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/grpc/grpc/12587)*
<!-- Sent from Reviewable.io -->

markdroth commented:
Reviewed 1 of 1 files at r7.
Review status: all files reviewed at latest revision, 9 unresolved discussions, some commit checks broke.

---

*[src/core/ext/filters/client_channel/lb_policy/grpclb/grpclb.c, line 1761 at r6](https://reviewable.io:443/reviews/grpc/grpc/12587#-KuyfseBp8Uhl5-1W9CS:-KuzqGM-2S7a6jCdnX1k:b-rybco9) ([raw file](https://github.com/grpc/grpc/blob/373d0783d37946586c4b127032191fa2a93f98b4/src/core/ext/filters/client_channel/lb_policy/grpclb/grpclb.c#L1761)):*
<details><summary><i>Previously, AspirinSJL (Juanli Shen) wrote…</i></summary><blockquote>

The problem of the previous code is that it doesn't include the case when `glb_policy->rr_policy == NULL` and we do want to propagate the update. For example, the fallback backends are restarted so there is a period without valid `rr_policy`. We saw this situation in our integration test.
</blockquote></details>

Ah, so this happens in the case where the RR policy shuts down?  Okay, then this is fine for now.  Eventually, we will want to change the RR policy not to shut down when all of its subchannels go bad, in which case this change would not be necessary.  But this alternative expression should work fine.

---

*[src/core/ext/filters/client_channel/lb_policy/grpclb/grpclb.c, line 1776 at r6](https://reviewable.io:443/reviews/grpc/grpc/12587#-KuyfaT_ZdnL2vQEBsZt:-KuzshgCr7u0xJfLuX8l:b-s8lpms) ([raw file](https://github.com/grpc/grpc/blob/373d0783d37946586c4b127032191fa2a93f98b4/src/core/ext/filters/client_channel/lb_policy/grpclb/grpclb.c#L1776)):*
<details><summary><i>Previously, AspirinSJL (Juanli Shen) wrote…</i></summary><blockquote>

Done.

BTW, `&& glb_policy->serverlist != NULL` is added in this line to overwrite the bogus balancer update which takes forever in the integration test.
</blockquote></details>

Hmm... This is interesting.  I'm actually a little surprised that this works, because in principle it is not necessarily safe to abort an in-progress update.  But I don't see anything in the code that prevents this.

Actually, looking at the current code, I'm not sure why we bother to track pending updates here at all.  We should be able to just send the updated addresses to the LB channel via the fake resolver unconditionally, and let the pick_first code in the LB channel handle the updates itself.  But we can make that change in a separate PR, so let's not worry about it for now.

---

*[src/core/ext/filters/client_channel/lb_policy/grpclb/grpclb.c, line 1785 at r7](https://reviewable.io:443/reviews/grpc/grpc/12587#-KuzsXHtvj9YQ147hae_:-KuzsXHtvj9YQ147haea:btny72m) ([raw file](https://github.com/grpc/grpc/blob/b2300febd8101fa2230bd561b59f7cd39eaf8c98/src/core/ext/filters/client_channel/lb_policy/grpclb/grpclb.c#L1785)):*
> ```C
>     // If a non-empty serverlist hasn't been received from the balancer,
>     // propagate
>     // the update to fallback_backend_addresses.
> ```

Nit: This can move to the previous line.

---

*[test/cpp/end2end/grpclb_end2end_test.cc, line 684 at r6](https://reviewable.io:443/reviews/grpc/grpc/12587#-Kuyi-ESxLjY6IC6VZl1:-Kuzsv0sUc3rhjqPAkow:b1eecxx) ([raw file](https://github.com/grpc/grpc/blob/373d0783d37946586c4b127032191fa2a93f98b4/test/cpp/end2end/grpclb_end2end_test.cc#L684)):*
<details><summary><i>Previously, AspirinSJL (Juanli Shen) wrote…</i></summary><blockquote>

I think we should use the previous comments, because `WaitForBackend()` works in a customer view instead of infrastructure view. If `WaitForBackend(i)` finishes, it only means backend `i` is reachable (from the client).
</blockquote></details>

What I mean here is that the goal of this loop is to ensure that all of the backends that we should have (not all of the entries in `backend_servers_`) are actually ready.  If the initial address list contains more than one backend and we check only the first of them, then we know that the address list was processed but we don't know that all of the backends are actually connected.  What we are checking here is both (a) that the address list is being used and (b) that the backends are connected.  The comment should reflect that.

---

*[test/cpp/end2end/grpclb_end2end_test.cc, line 759 at r6](https://reviewable.io:443/reviews/grpc/grpc/12587#-Kuyi4F1f_vGbNo-JiFc:-KuztZvRP2AX3C3FPPKS:b-q64pkt) ([raw file](https://github.com/grpc/grpc/blob/373d0783d37946586c4b127032191fa2a93f98b4/test/cpp/end2end/grpclb_end2end_test.cc#L759)):*
<details><summary><i>Previously, AspirinSJL (Juanli Shen) wrote…</i></summary><blockquote>

The same as above (line 684).
</blockquote></details>

See above.

---

*[test/cpp/end2end/grpclb_end2end_test.cc, line 784 at r6](https://reviewable.io:443/reviews/grpc/grpc/12587#-Kuyi6x-tTlRKk7RKtP-:-Kuzt_tFjZzSfjEFZdLq:b-q64pkt) ([raw file](https://github.com/grpc/grpc/blob/373d0783d37946586c4b127032191fa2a93f98b4/test/cpp/end2end/grpclb_end2end_test.cc#L784)):*
<details><summary><i>Previously, AspirinSJL (Juanli Shen) wrote…</i></summary><blockquote>

The same as above (line 684).
</blockquote></details>

See above.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/grpc/grpc/12587)*
<!-- Sent from Reviewable.io -->

AspirinSJL commented:
Review status: all files reviewed at latest revision, 7 unresolved discussions, some commit checks broke.

---

*[src/core/ext/filters/client_channel/lb_policy/grpclb/grpclb.c, line 1761 at r6](https://reviewable.io:443/reviews/grpc/grpc/12587#-KuyfseBp8Uhl5-1W9CS:-Kv-AXwlhB_pcrQS1Bd9:btdsfzr) ([raw file](https://github.com/grpc/grpc/blob/373d0783d37946586c4b127032191fa2a93f98b4/src/core/ext/filters/client_channel/lb_policy/grpclb/grpclb.c#L1761)):*
<details><summary><i>Previously, markdroth (Mark D. Roth) wrote…</i></summary><blockquote>

Ah, so this happens in the case where the RR policy shuts down?  Okay, then this is fine for now.  Eventually, we will want to change the RR policy not to shut down when all of its subchannels go bad, in which case this change would not be necessary.  But this alternative expression should work fine.
</blockquote></details>

I see.

---

*[src/core/ext/filters/client_channel/lb_policy/grpclb/grpclb.c, line 1776 at r6](https://reviewable.io:443/reviews/grpc/grpc/12587#-KuyfaT_ZdnL2vQEBsZt:-Kv-BYBTvr0EqEZvvwc_:b-vbdp0r) ([raw file](https://github.com/grpc/grpc/blob/373d0783d37946586c4b127032191fa2a93f98b4/src/core/ext/filters/client_channel/lb_policy/grpclb/grpclb.c#L1776)):*
<details><summary><i>Previously, markdroth (Mark D. Roth) wrote…</i></summary><blockquote>

Hmm... This is interesting.  I'm actually a little surprised that this works, because in principle it is not necessarily safe to abort an in-progress update.  But I don't see anything in the code that prevents this.

Actually, looking at the current code, I'm not sure why we bother to track pending updates here at all.  We should be able to just send the updated addresses to the LB channel via the fake resolver unconditionally, and let the pick_first code in the LB channel handle the updates itself.  But we can make that change in a separate PR, so let's not worry about it for now.
</blockquote></details>

Yes, the PF policy should be able to handle this itself.

---

*[src/core/ext/filters/client_channel/lb_policy/grpclb/grpclb.c, line 1785 at r7](https://reviewable.io:443/reviews/grpc/grpc/12587#-KuzsXHtvj9YQ147hae_:-Kv-C0urSJ3GN52xzxiF:b-urihqg) ([raw file](https://github.com/grpc/grpc/blob/b2300febd8101fa2230bd561b59f7cd39eaf8c98/src/core/ext/filters/client_channel/lb_policy/grpclb/grpclb.c#L1785)):*
<details><summary><i>Previously, markdroth (Mark D. Roth) wrote…</i></summary><blockquote>

Nit: This can move to the previous line.
</blockquote></details>

Done. Sorry about that.

---

*[test/cpp/end2end/grpclb_end2end_test.cc, line 684 at r6](https://reviewable.io:443/reviews/grpc/grpc/12587#-Kuyi-ESxLjY6IC6VZl1:-Kv-CPVKkXGr-vzC6Quz:b-b9hsr) ([raw file](https://github.com/grpc/grpc/blob/373d0783d37946586c4b127032191fa2a93f98b4/test/cpp/end2end/grpclb_end2end_test.cc#L684)):*
<details><summary><i>Previously, markdroth (Mark D. Roth) wrote…</i></summary><blockquote>

What I mean here is that the goal of this loop is to ensure that all of the backends that we should have (not all of the entries in `backend_servers_`) are actually ready.  If the initial address list contains more than one backend and we check only the first of them, then we know that the address list was processed but we don't know that all of the backends are actually connected.  What we are checking here is both (a) that the address list is being used and (b) that the backends are connected.  The comment should reflect that.
</blockquote></details>

Ah, I see. That's also why I changed the previous code into a for loop over each subset. 

Changed the comment.

---

*[test/cpp/end2end/grpclb_end2end_test.cc, line 759 at r6](https://reviewable.io:443/reviews/grpc/grpc/12587#-Kuyi4F1f_vGbNo-JiFc:-Kv-FRRUWGPNU_cqJwYU:b-896fix) ([raw file](https://github.com/grpc/grpc/blob/373d0783d37946586c4b127032191fa2a93f98b4/test/cpp/end2end/grpclb_end2end_test.cc#L759)):*
<details><summary><i>Previously, markdroth (Mark D. Roth) wrote…</i></summary><blockquote>

See above.
</blockquote></details>

Done.

---

*[test/cpp/end2end/grpclb_end2end_test.cc, line 784 at r6](https://reviewable.io:443/reviews/grpc/grpc/12587#-Kuyi6x-tTlRKk7RKtP-:-Kv-FSG3bNaEEe3Bp1zM:b-896fix) ([raw file](https://github.com/grpc/grpc/blob/373d0783d37946586c4b127032191fa2a93f98b4/test/cpp/end2end/grpclb_end2end_test.cc#L784)):*
<details><summary><i>Previously, markdroth (Mark D. Roth) wrote…</i></summary><blockquote>

See above.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/grpc/grpc/12587)*
<!-- Sent from Reviewable.io -->

grpc-jenkins commented:
```
[trickle] No significant performance differences
```

AspirinSJL commented:
I filed the "issue" about propagating update in #12727.

grpc-jenkins commented:
```
Corrupt JSON data (indicates timeout or crash): 
    bm_fullstack_trickle.BM_PumpStreamServerToClient_Trickle_2M_32k.opt.new: 1


[trickle] No significant performance differences
```

grpc-jenkins commented:
```
Corrupt JSON data (indicates timeout or crash): 
    bm_fullstack_streaming_ping_pong.BM_StreamingPingPongWithCoalescingApi_MinInProcessCHTTP2_NoOpMutator_NoOpMutator__2M_2_0.opt.new: 1


[microbenchmarks] Performance differences noted:
Benchmark                                                                                                  allocs_per_iteration    atm_add_per_iteration    cpu_time    nows_per_iteration    real_time
---------------------------------------------------------------------------------------------------------  ----------------------  -----------------------  ----------  --------------------  -----------
BM_HpackEncoderEncodeHeader<SingleInternedBinaryElem<1, true>>/0/16k                                                                                        +4%                               +4%
BM_HpackEncoderEncodeHeader<SingleInternedBinaryElem<100, false>>/0/16k                                                                                     +9%                               +9%
BM_HpackEncoderEncodeHeader<SingleNonInternedElem>/0/16k                                                                                                    +4%                               +5%
BM_HpackParserParseHeader<RepresentativeServerInitialMetadata>                                                                                                                                +4%
BM_PumpStreamServerToClient<InProcess>/64                                                                                                                   +4%                               +4%
BM_PumpStreamServerToClient<TCP>/256k                                                                                                                                                         +5%
BM_StreamingPingPong<InProcess, NoOpMutator, NoOpMutator>/8/2                                                                                               +16%                              +17%
BM_StreamingPingPong<MinTCP, NoOpMutator, NoOpMutator>/512/1                                                                                                +6%                               +6%
BM_StreamingPingPongWithCoalescingApi<InProcess, NoOpMutator, NoOpMutator>/512/2/0                                                                          +6%                               +6%
BM_StreamingPingPongWithCoalescingApi<InProcessCHTTP2, NoOpMutator, NoOpMutator>/1/1/1                                                                      +4%
BM_StreamingPingPongWithCoalescingApi<MinInProcess, NoOpMutator, NoOpMutator>/64/2/1                                                                        +4%                               +4%
BM_StreamingPingPongWithCoalescingApi<MinInProcessCHTTP2, NoOpMutator, NoOpMutator>/4k/1/1                                                                  +8%                               +8%
BM_TransportStreamSend/32k                                                                                                                                  +13%                              +13%
BM_UnaryPingPong<InProcess, Client_AddMetadata<RandomAsciiMetadata<10>, 1>, NoOpMutator>/0/0                                                                +7%                               +7%
BM_UnaryPingPong<InProcess, Client_AddMetadata<RandomAsciiMetadata<31>, 1>, NoOpMutator>/0/0                                                                +5%                               +5%
BM_UnaryPingPong<InProcess, NoOpMutator, Server_AddInitialMetadata<RandomBinaryMetadata<100>, 1>>/0/0                                                       +8%                               +8%
BM_UnaryPingPong<InProcessCHTTP2, NoOpMutator, NoOpMutator>/64/0                                                                                            +5%                               +5%
BM_UnaryPingPong<InProcessCHTTP2, NoOpMutator, Server_AddInitialMetadata<RandomAsciiMetadata<31>, 1>>/0/0                                                   +8%                               +8%
BM_UnaryPingPong<MinInProcess, NoOpMutator, NoOpMutator>/0/512                                                                                              +5%                               +5%
BM_UnaryPingPong<MinInProcessCHTTP2, NoOpMutator, NoOpMutator>/0/16M                                                                                                    -6%
BM_UnaryPingPong<MinInProcessCHTTP2, NoOpMutator, NoOpMutator>/512/512                                                                                      +4%                               +4%
BM_UnaryPingPong<MinTCP, NoOpMutator, NoOpMutator>/2M/2M                                                                           -4%                                  -4%
BM_UnaryPingPong<TCP, NoOpMutator, NoOpMutator>/128M/128M                                                  -11%                    -10%
```

AspirinSJL commented:
Basic Tests C++ Linux [dbg] (internal CI): #12510
Basic Tests C++ Linux [opt] (internal CI): #12313
Basic Tests MacOS [dbg] (internal CI): #12138
Basic Tests MacOS [opt] (internal CI): #12138
Basic Tests Windows [dbg] (internal CI): #12595
Basic Tests Windows [opt] (internal CI): #12595
Interop Cloud-to-Cloud Tests (internal CI): #12723
Interop Cloud-to-Prod Tests (internal CI): #12260
Tsan C (internal CI) — Internal CI build failed: #12741

grpc-jenkins commented:
```
[microbenchmarks] No significant performance differences
```
